File: Leetcode Questions/977-Square-of-Sorted-Array/977.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;


vector<int> sortedSquares(vector<int> &arr) {
    int n = arr.size(), sep = -1;
    vector<int> final;

    for(int i=0; i<n; i++) {
        if(arr[i] >= 0) {
            sep = i;
            break;
        }
    }

    if(sep == -1) {
        reverse(arr.begin(), arr.end());
        for(int i: arr)
            final.push_back(i*i);
        return final;
    }

    int left = sep - 1, right = sep;
    while(left >= 0 && right < n) {
        if(arr[right] < abs(arr[left])) {
            final.push_back(arr[right]*arr[right]);
            right++;
        }
        else {
            final.push_back(arr[left]*arr[left]);
            left--;
        }
    }

    while(left >= 0) {
        final.push_back(arr[left]*arr[left]);
        left--;
    }

    while(right < n) {
        final.push_back(arr[right]*arr[right]);
        right++;
    }

    return final;
}




int main() {
    vector<int> nums = {-4,-1,0,3,10};
    vector<int> nums1 = {-7,-3,2,3,11};
    // vector<int> ans = sortedSquares(nums);
    cout<<"Sorted Squares: "<<endl;
    for(int i: sortedSquares(nums)) {
        cout<<i<<" ";
    }
    return 0;
}

File: Leetcode Questions/977-Square-of-Sorted-Array/977.py
'''
Given an integer array nums sorted in non-decreasing order, return an array of the squares of each number sorted in non-decreasing order.

Example 1:
Input: nums = [-4,-1,0,3,10]
Output: [0,1,9,16,100]
Explanation: After squaring, the array becomes [16,1,0,9,100].
After sorting, it becomes [0,1,9,16,100].
'''


# Brute Force: square all elements and then sort. TC: O(nlog(n)), SC: O(1)

# Optimal: My Intuition: find the separation(sep) index from where the positive numbers start, maintain two pointers;
# one to the left of sep and another to the right of sep. Iterate the left and right pointers while comparing the elements,
# squaring and appending them to new list/array
# TC: O(n), SC: O(n)

def sortedSquares(arr: list[int]) -> list[int]:
    n = len(arr)
    sep = -1
    final = []
    
    for i in range(n):
        if arr[i] >= 0:
            sep = i
            break
    
    if sep == -1:          # ie all negative numbers in given array
        arr.reverse()
        final = [x**2 for x in arr]
        return final
    
    left = sep - 1
    right = sep
    
    while left >=0 and right < n:
        if arr[right] < abs(arr[left]):
            final.append((arr[right])**2)
            right += 1
        else:
            final.append((arr[left])**2)
            left -= 1
    
    while left >= 0:                    # to check if any elements are left or not
        final.append((arr[left])**2)
        left -= 1
    
    while right < n:                    # to check if any elements are left or not
        final.append((arr[right])**2)
        right += 1
        
    return final
    
    
    
# Making use of two poiniters, append from the back of the final array ie checking for greater than condition. Another solution
def sortedSquares1(arr: list[int]) -> list[int]:
    n = len(arr)
    final = [0]*n
    start, end = 0, n-1
    for i in range(n-1, -1, -1):
        if abs(arr[start]) >= abs(arr[end]):
            final[i] = arr[start]**2
            start += 1
        else:
            final[i] = arr[end]**2
            end -= 1
    return final
    
    
    
    
if __name__ == "__main__":
    nums = [-4,-1,0,3,10]
    nums1 = [-7,-3,2,3,11]
    print(sortedSquares1(nums1))